### ISSUE TITLE
Push Prototype Step Crashes with ResNet50 Configuration (width_index >= n Error)

### ISSUE THREAD

**Issue by KatiaBenali** [cwangrun/ST-ProtoPNet#1]
I encountered an issue when running the ST-ProtoPNet code with a ResNet50 backbone (pre-trained on iNaturalist). The model works correctly with the VGG16 configuration, but with ResNet50, it crashes during the "push prototype" step. The error occurs because width_index exceeds n, suggesting an out-of-bounds issue during feature extraction.

**Configuration :**
Backbone: ResNet50
Pretrained weights: iNaturalist-pretrained ResNet50
Image size: 224x224 (default)
Number of prototypes: 5 per class (as defined in the config file).

**The code segment**
The issue appears in the following segment of receptive_field.py : 
```
def compute_rf_protoL_at_spatial_location(img_size, height_index, width_index, protoL_rf_info):
    n = protoL_rf_info[0]
    j = protoL_rf_info[1]
    r = protoL_rf_info[2]
    start = protoL_rf_info[3]
    assert(height_index < n)
    assert(width_index < n)
```
and in push_trivial.py
`for j in range(n_prototypes):  # 980 prototypes
    if class_specific:
        target_class = torch.argmax(prototype_network_parallel.module.prototype_class_identity[j]).item()
        if len(class_to_img_index_dict[target_class]) == 0:
            continue
        proto_dist_j = proto_dist_[class_to_img_index_dict[target_class]][:, j, :, :]
    else:
        proto_dist_j = proto_dist_[:, j, :, :]  # Shape: [batch_size, 14, 14]

    batch_min_proto_dist_j = np.amin(proto_dist_j)
    if batch_min_proto_dist_j < global_min_proto_dist[j]:
        batch_argmin_proto_dist_j = list(np.unravel_index(np.argmin(proto_dist_j, axis=None), proto_dist_j.shape))
        if class_specific:
            batch_argmin_proto_dist_j[0] = class_to_img_index_dict[target_class][batch_argmin_proto_dist_j[0]]

        # Problematic calculation
        fmap_width_start_index = batch_argmin_proto_dist_j[2] * prototype_layer_stride
        fmap_width_end_index = fmap_width_start_index + proto_w
`
**Observed behaviour :**
When using ResNet50, width_index (i.e., batch_argmin_proto_dist_j[2]) exceeds n, causing an out-of-bounds error.
The feature map shape is (1, 2048, 7, 7), so n = 7. However, width_index sometimes equals or exceeds 8, which causes the crash.

**Expected Behavior:**
The width_index should always be within the range [0, n-1]. The prototype push process should correctly extract feature patches without exceeding the feature map dimensions.

**Question :** Are there specific adjustments required for the receptive field calculation when using ResNet50?
Any suggestions on how to resolve this problem ?

Thank you in advance for your assistance!

**Comment by cwangrun** [comment]
Hi, thanks for the question. This is because the additional up-sampling step in the add-on layers makes the feature size become 14*14, rather than 7*7. I think it can be fixed by resetting the actual feature size and stride, please see the updated model file.
